### PR TITLE
[NVIDIA XLA:GPU] Fused MHA - Infrastructure + BMM-BMM fusion

### PR DESCRIPTION
**Pull request by AyanmoI**
cuDNN currently supports and the rewriter matches the following patterns for Multi-headed attention:

BMM1 - BMM2
BMM1 - Scale - Bias - Mask - Softmax - BMM2
BMM1 - Scale - Bias - Mask - Softmax - Dropout - BMM2
BMM1 - Scale - Mask - Softmax - BMM2
BMM1 - Scale - Mask - Softmax - Dropout - BMM2
BMM1 - Softmax - Dropout - BMM2
The following pattern is matched here:
`Output = BMM2(BMM1(Q,K), V)` 
and the infrastructure to support the other patterns

### CODE REVIEW

**Comment by gbaned** [comment]
@AyanmoI Can you please resolve conflicts? Thank you!

File: tensorflow/compiler/xla/service/gpu/backend_configs.proto [hunk 1]
```diff
@@ -88,3 +88,22 @@ message BitcastBackendConfig {
   LayoutProto source_layout = 1;
   LayoutProto result_layout = 2;
 }
+
+// Backend config for a fused Multi-Headed Attention (fMHA) that runs through
+// cudnn.
+message CudnnfMHABackendConfig {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/backend_configs.proto:94]: Maybe just MHABackendConfig, given that names in this file are generally library-agnostic?

File: tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc [hunk 2]
```diff
@@ -107,5 +133,40 @@ std::string CudnnConvKindToString(CudnnConvKind kind) {
   }
 }
 
+StatusOr<CudnnfMHAKind> GetCudnnfMHAKind(
+    const HloCustomCallInstruction* instr) {
+  absl::string_view target = instr->custom_call_target();
+  if (target == kCudnnfMHABmmBmmCallTarget) return CudnnfMHAKind::kBmmBmm;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc:139]: Nit: braceless ifs are generally avoided in the codebase, as adding logging statements can lead to goto-fails.

File: tensorflow/compiler/xla/service/gpu/cublas_cudnn.h [hunk 3]
```diff
@@ -126,6 +135,32 @@ bool IsCustomCallToDnnConvolution(const HloInstruction& hlo);
 // reordering helper (required for int8x32 convolutions).
 bool IsCudnnConvolutionReorder(const HloInstruction& hlo);
 
+// The fused_mha_rewriter phase where each of the MHA signatures are pattern
+// matched and rewritten into a custom-call with specific custom-call target.
+// The custom-call target specifies the MHA signature. For example,  BMM1 - Bias
+// - Scale - Mask - Softmax - BMM2 pattern can have the target as
+// cudnn$fmhaBiasScaleMaskSoftmax.
+// The fMHA signatures currently supported by cudnn are:
+// 1. BMM1 - BMM2
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cublas_cudnn.h:144]: Maybe attach each comment to each string_view? Otherwise those can get out of order.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+bool IsBatchedMatmul(const HloInstruction* gemm) {
+  return IsCublasGemm(*gemm) && IsBatchedDot(gemm).value();
+}
+
+// Give this instruction a more useful name than "custom-call.42".
+Status SetName(HloModule* module, HloInstruction* fmha) {
+  if (fmha->custom_call_target() == kCudnnfMHABmmBmmCallTarget) {
+    module->SetAndUniquifyInstrName(fmha, "fmha-bmm-bmm");
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:69]: Nit: why not just use the existing string constant?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+  return InternalError(
+      "Found invalid FMHA custom-call target while setting custom-call name");
+}
+
+bool IsComputeCapabilitySupported(stream_executor::CudaComputeCapability cc) {
+  if (!(cc.IsAtLeast(
+          se::CudaComputeCapability::AMPERE)  && cc.minor == 0 )) {
+      VLOG(2) << "CudnnFusedMHARewriter did not run. Unsupported compute "
+                 "capability. Only GA100 supported.";
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:106]: Nit: it's not GA100, it's compute capability? Should it also check for cuDNN version?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str, GetModuleConfig()));
+  CudnnFusedMHARewriter fusedMhaRewriter{GetCudaComputeCapability()};
+  TF_ASSERT_OK_AND_ASSIGN(bool result, RunHloPass(&fusedMhaRewriter, m.get()));
+  EXPECT_FALSE(result);
+}
+
+TEST_F(CudnnFusedMhaRewriterTestHloTest, F16Bmm1Bmm2Pattern) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:193]: Presumably we would need more tests for more fMHA variants?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 7]
```diff
@@ -1204,6 +1232,156 @@ Status IrEmitterUnnested::EmitConvolutionReorderThunk(mlir::Operation* op) {
...
+    return OkStatus();
+  };
+
+  enum OptionalBuffer {
+    kBias,
+    kMask,
+  };
+  BufferAllocation::Slice mask_slice;
+  BufferAllocation::Slice bias_slice;
+  auto allocate_buffer = [&](auto fmha, OptionalBuffer optional_buffer_kind) {
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:1316]: The name seems misleading: the function is not allocating anything, right?

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc [hunk 8]
```diff
@@ -4046,8 +4047,193 @@ GetCudnnFusedMatmulGraph(dnn::DataType input_type, dnn::DataType bias_type,
   return std::make_unique<cudnn_frontend::OperationGraph>(std::move(op_graph));
 }
 
+tsl::StatusOr<std::unique_ptr<cudnn_frontend::OperationGraph>>
+GetCudnnFusedMHASimpleOperationGraph(
```
Comment (cheshire) [inline tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: In terms of slicing PRs, MHA support in SE could probably land by itself (is there a way to e2e test it just in SE?)

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc [hunk 9]
```diff
@@ -4046,8 +4047,193 @@ GetCudnnFusedMatmulGraph(dnn::DataType input_type, dnn::DataType bias_type,
...
+          << "\nBMM2: " << bmm2_desc.describe()
+          << "\nBMM2_op: " << bmm2_op.describe()
+          << "\nOpGraph: " << op_graph.describe();
+  return std::unique_ptr<cudnn_frontend::OperationGraph>(
+      new cudnn_frontend::OperationGraph(std::move(op_graph)));
+}
+
 }  // namespace
 
+// // This method uses cudnn_frontend::cudnnException which is currently not
```
Comment (cheshire) [inline tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4192]: Remove commented out code?

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc [hunk 10]
```diff
@@ -5481,6 +5667,80 @@ bool CudnnSupport::GetConvolveAlgorithms(
...
+  TF_ASSIGN_OR_RETURN(auto op_graph,
+                      GetCudnnFusedMHASimpleOperationGraph(
+                          kind, bmm1_lhs_descriptor, bmm1_rhs_descriptor,
+                          bmm2_rhs_descriptor, intermediate_bmm2_lhs_descriptor,
+                          output_descriptor, dropout_rate, cudnn));
+  // The function  GetExecPlanFromHeuristics uses cudnn_frontend::cudnnException
+  // which is currently not
+  // recommended for use by Google. Hence commenting out the call to this
+  // function until this gets resolved.
+  // TF_ASSIGN_OR_RETURN(auto execution_plan,
```
Comment (cheshire) [inline tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:5691]: We do not check in commented out code. If you need exception support, could you add wrappers which will be compiled with their own rules, and convert exceptions to Status?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/backend_configs.proto (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/backend_configs.proto:94]: Done.

File: tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc (hunk 2 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cublas_cudnn.cc:139]: Done.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 4 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:69]: Which string constant exactly? the custom-call target string?

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 5 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc:106]: It's actually GA100. Hence checking the minor as well. I will add code to the cudnn version in the next commit.

File: tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc (hunk 6 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:193]: Yes. We will have the tests along with support for those patterns.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 7 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:1316]: Changed to `get_buffer_allocation_slice`. Done.

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc (hunk 8 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: Unfortunately no. There is no way that I can think of that we test SE changes only. Do u have any suggestions?

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc (hunk 9 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4192]: Yeah. This was an internal thing. Removed.

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc (hunk 10 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:5691]: We plan to add some support for wrappers for `cudnn_frontend::cudnnException`. However, that doesn't seem to a high priority right now since we have a workaround to use knob ids n values. Removing the commented code.

**Comment by AyanmoI** [comment]
> @AyanmoI Can you please resolve conflicts? Thank you!

Rebased and resolved conflicts.

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc (hunk 8 above)
Comment (cheshire) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: OK indeed we don't really test things on SE level (probably an oversight), but then I think it's lesser evil to commit SE stuff first (even without tests =/)

Comment (AyanmoI) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: In that case, we can create a PR with SE changes for all the patterns that we intend to have?

Comment (AyanmoI) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: @cheshire Any comments on the above?

Comment (cheshire) [reply tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:4051]: Yes please.

**Comment by AyanmoI** [comment]
This will be spilt into 2 PRs, one of which is https://github.com/tensorflow/tensorflow/pull/60047. Closing this one.
